sql: clean up and simplify scan-related code a bit

yuzefovich opened the pull request:
**sql,colexec: minor cleanup of scan-related things**

This commit removes multiple arguments in favor of using the spec when
initializing the cFetcher as well as refactors several loops in order to
avoid making copies of descriptions during table reader planning.
Additionally, it moves a map from scanNode next to the map's only user
(stats job) and removes the map from the scanNode itself.

Release note: None

**sql: remove ScanVisibility in favor of using PublicAndNonPublic**

Several years ago we introduced ScanVisibility to let the disk-reading
processors know whether they should be looking only at "public" or
"public and non-public" columns (the latter include the columns under
active schema changes). This distinction is likely not that useful
anymore, so this commit removes the scan visibility in favor of using
PublicAndNonPublic throughout the codebase. This was already the case
for the table readers and join readers, so the change affects the
inverted joiners and zigzag joiners.

Release note: None

yuzefovich committed:
sql,colexec: minor cleanup of scan-related things

This commit removes multiple arguments in favor of using the spec when
initializing the cFetcher as well as refactors several loops in order to
avoid making copies of descriptions during table reader planning.
Additionally, it moves a map from scanNode next to the map's only user
(stats job) and removes the map from the scanNode itself.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56537)
<!-- Reviewable:end -->

yuzefovich commented:
Filter benchmark of four commits:
```
name                                           old time/op    new time/op    delta
FlowSetup/vectorize=true/distribute=true-24      19.0µs ± 3%    18.7µs ± 5%  -1.42%  (p=0.009 n=22+24)
FlowSetup/vectorize=true/distribute=false-24     16.9µs ± 2%    16.6µs ± 3%  -1.69%  (p=0.000 n=25+24)
FlowSetup/vectorize=false/distribute=true-24     18.3µs ± 3%    18.1µs ± 2%  -1.16%  (p=0.007 n=25+25)
FlowSetup/vectorize=false/distribute=false-24    16.4µs ± 5%    16.3µs ± 4%    ~     (p=0.162 n=25+25)

name                                           old alloc/op   new alloc/op   delta
FlowSetup/vectorize=true/distribute=true-24      33.8kB ± 0%    32.6kB ± 0%  -3.74%  (p=0.000 n=20+22)
FlowSetup/vectorize=true/distribute=false-24     30.4kB ± 0%    29.1kB ± 0%  -4.28%  (p=0.000 n=25+25)
FlowSetup/vectorize=false/distribute=true-24     34.3kB ± 0%    34.2kB ± 0%  -0.47%  (p=0.000 n=22+22)
FlowSetup/vectorize=false/distribute=false-24    30.9kB ± 0%    30.7kB ± 0%  -0.51%  (p=0.000 n=22+21)

name                                           old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true-24         323 ± 0%       311 ± 0%  -3.72%  (p=0.000 n=25+24)
FlowSetup/vectorize=true/distribute=false-24        294 ± 1%       282 ± 1%  -4.09%  (p=0.000 n=25+25)
FlowSetup/vectorize=false/distribute=true-24        288 ± 0%       287 ± 0%  -0.35%  (p=0.000 n=25+25)
FlowSetup/vectorize=false/distribute=false-24       257 ± 0%       256 ± 0%  -0.20%  (p=0.000 n=25+25)
```
KV benchmark of four commits:
```
name                                           old time/op    new time/op    delta
FlowSetup/vectorize=true/distribute=true-24      18.3µs ± 3%    18.2µs ± 5%    ~     (p=0.088 n=22+24)
FlowSetup/vectorize=true/distribute=false-24     16.4µs ± 3%    16.2µs ± 3%  -1.25%  (p=0.008 n=25+25)
FlowSetup/vectorize=false/distribute=true-24     17.5µs ± 3%    17.3µs ± 3%    ~     (p=0.058 n=24+25)
FlowSetup/vectorize=false/distribute=false-24    15.6µs ± 5%    15.5µs ± 3%    ~     (p=0.140 n=24+24)

name                                           old alloc/op   new alloc/op   delta
FlowSetup/vectorize=true/distribute=true-24      32.6kB ± 0%    32.0kB ± 0%  -1.85%  (p=0.000 n=23+23)
FlowSetup/vectorize=true/distribute=false-24     29.2kB ± 0%    28.6kB ± 0%  -2.09%  (p=0.000 n=25+25)
FlowSetup/vectorize=false/distribute=true-24     33.0kB ± 0%    32.8kB ± 0%  -0.47%  (p=0.000 n=23+22)
FlowSetup/vectorize=false/distribute=false-24    29.6kB ± 0%    29.4kB ± 0%  -0.52%  (p=0.000 n=23+20)

name                                           old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true-24         311 ± 0%       302 ± 0%  -2.89%  (p=0.000 n=24+24)
FlowSetup/vectorize=true/distribute=false-24        285 ± 1%       276 ± 1%  -3.16%  (p=0.000 n=25+25)
FlowSetup/vectorize=false/distribute=true-24        270 ± 0%       269 ± 0%  -0.37%  (p=0.000 n=25+25)
FlowSetup/vectorize=false/distribute=false-24       242 ± 0%       241 ± 0%  -0.41%  (p=0.000 n=25+20)
```

yuzefovich committed:
sql: remove ScanVisibility in favor of using PublicAndNonPublic

Several years ago we introduced ScanVisibility to let the disk-reading
processors know whether they should be looking only at "public" or
"public and non-public" columns (the latter include the columns under
active schema changes). This distinction is likely not that useful
anymore, so this commit removes the scan visibility in favor of using
PublicAndNonPublic throughout the codebase. This was already the case
for the table readers and join readers, so the change affects the
inverted joiners and zigzag joiners.

Note that I didn't bump the distsql version because we've just bumped it
in another commit.

Release note: None

yuzefovich commented:
First commit has been merged, and I no longer think that the second commit is reasonable, so closing.
